azure_functions_worker/extension.py

```diff
@@ -0,0 +1,297 @@
+        hook_name: str
+            The exetension name to be executed (e.g. pre_invocations).
+            These are defined in azure.functions.FuncExtensionHooks.
+        """
+        sdk = cls._try_get_sdk_with_extension_enabled()
+        if sdk is None:
+            return
+
+        # Get function hooks from azure.functions.extension.ExtensionMeta
+        # The return type is FuncExtensionHooks
+        funcs = sdk.ExtensionMeta.get_function_hooks(ctx.function_name)
+
+        # Invoke function hooks
+        cls._safe_execute_invocation_hooks(
+            funcs, hook_name, ctx, func_args, func_ret
+        )
+
+        # Get application hooks from azure.functions.extension.ExtensionMeta
+        # The reutnr type is AppExtensionHooks
+        apps = sdk.ExtensionMeta.get_application_hooks()
+
+        # Invoke application hook
+        cls._safe_execute_invocation_hooks(
+            apps, hook_name, ctx, func_args, func_ret
+        )
+
+    @classmethod
+    def get_sync_invocation_wrapper(cls, ctx, func) -> Callable[[List], Any]:
+        """Get a synchronous lambda of extension wrapped function which takes
+        function parameters
+        """
+        return functools.partial(cls._raw_invocation_wrapper, ctx, func)
+
+    @classmethod
+    async def get_async_invocation_wrapper(cls, ctx, function, args) -> Any:
+        """An asynchronous coroutine for executing function with extensions
+        """
+        cls._invocation_extension(ctx, APP_EXT_PRE_INVOCATION, args)
+        cls._invocation_extension(ctx, FUNC_EXT_PRE_INVOCATION, args)
+        result = await function(**args)
+        cls._invocation_extension(ctx, FUNC_EXT_POST_INVOCATION, args, result)
+        cls._invocation_extension(ctx, APP_EXT_POST_INVOCATION, args, result)
+        return result
+
+    @staticmethod
+    def _get_sdk_from_sys_path() -> ModuleType:
+        """Get the azure.functions SDK from the latest sys.path defined.
+        This is to ensure the extension loaded from SDK is actually coming from
+        customer's site-packages.
+
+        Returns
+        -------
+        ModuleType
+            The azure.functions that is loaded from the first sys.path entry
+        """
+        if 'azure.functions' in sys.modules:
+            sys.modules.pop('azure.functions')
+
+        return importlib.import_module('azure.functions')
+
+    @staticmethod
+    def _is_extension_enabled_in_sdk(module: ModuleType) -> bool:
+        """Check if the extension feature is enabled in particular
+        azure.functions package.
+
+        Parameters
+        ----------
+        module: ModuleType
+            The azure.functions SDK module
+
+        Returns
+        -------
+        bool
+            True on azure.functions SDK supports extension registration
+        """
+        return getattr(module, 'ExtensionMeta', None) is not None
+
+    @staticmethod
+    def _get_sdk_version(module: ModuleType) -> str:
```

Comment:
Same comment as above - this feels like something that should be outside of this specific class.